posts/2024-03-13-testcontainers-guide.adoc

```diff
@@ -0,0 +1,84 @@
+
+image::/img/blog/testcontainers_banner.png[Testcontainers logo,width=50%,align="center"]
+
+== Why is Testcontainers helpful?
+
+Although this technology is especially useful for applications that have external resource dependencies (such as databases, message queues, or web services), one of the most important features of Testcontainers is the generic support it has for any Docker image. By encapsulating any necessary dependencies in containers, Testcontainers simplifies the configuration process and ensures a uniform testing setup that closely mirrors production environments.
+
+image::/img/blog/TestContainers3.png[Diagram of generic Testcontainers architecture,width=50%,align="center"]
+
+In essence, with Testcontainers, you can run your tests in an isolated and controlled environment that closely resembles your production setup, ensuring that your tests are reliable and reproducible.
+
+As well as this increased reliability and reproducibility of tests, there are additional key features/benefits that make Testcontainers a really useful tool, including:
+
+* Offering a wide range of supported containers
+
++
+Testcontainers offers support for a large variety of different Docker containers, including databases (e.g., PostgreSQL, MySQL, MongoDB, DB2, Redis), messaging brokers (e.g., RabbitMQ, Kafka), web servers, and more. This diversity enables developers to be able to customise test environments to make them specific to their own application stack.
+
+* Integration with popular testing frameworks
+
++
+Testcontainers integrates with popular testing frameworks including JUnit, TestNG, and Spock. This enables developers to easily and efficiently incorporate Testcontainers into their testing suite without having to make significant changes to their existing tests.
+
+
+* Declarative configuration
+
++
+Testcontainers provides a simple, declarative API for configuring and deploying containers. Developers are able to define any desired properties of a container using code, making it easy to configure and maintain.
+
+
+* Lifecycle management
+
++
+Testcontainers also handles the lifecycle of containers, ensuring they are started before the tests run and, importantly, stopped afterward. This automation removes this responsibility from the shoulders of developers, eliminating the need for manual intervention and making it easy to maintain clean, isolated testing environments.
+
+
+* Wait strategies
+
++
+To prevent your tests from trying to run before any required containers are set up, Testcontainers has a helpful, built-in wait strategy. This is of particular importance for any services that have tests that rely on services that can take significant time to initialise (e.g. databases).
```

Comment:
```suggestion
To prevent your tests from trying to run before any required containers are set up, Testcontainers has a helpful, built-in wait strategy. This particularly important for services with tests that rely on services that can take significant time to initialise (e.g. databases).
```